python/lancedb/embeddings/functions.py

```diff
@@ -0,0 +1,183 @@
+
+        Parameters
+        ----------
+        metadata : dict
+            The metadata from an arrow table. Note that
+            the keys and values are bytes
+        """
+        serialized = metadata[b"embedding_functions"]
+        raw_list = json.loads(serialized.decode("utf-8"))
+        functions = {}
+        for obj in raw_list:
+            model = self.load(obj["schema"]["title"])
+            functions[obj["model"]["vector_column"]] = model(**obj["model"])
+        return functions
+
+    def function_to_metadata(self, func):
+        """
+        Convert the given embedding function and source / vector column configs
+        into a config dictionary that can be serialized into arrow metadata
+        """
+        schema = func.model_json_schema()
+        json_data = func.model_dump()
+        return {
+            "schema": schema,
+            "model": json_data,
+        }
+
+    def get_table_metadata(self, func_list):
+        """
+        Convert a list of embedding functions and source / vector column configs
+        into a config dictionary that can be serialized into arrow metadata
+        """
+        json_data = [self.function_to_metadata(func) for func in func_list]
+        # Note that metadata dictionary values must be bytes so we need to json dump then utf8 encode
+        metadata = json.dumps(json_data, indent=2).encode("utf-8")
+        return {"embedding_functions": metadata}
+
+
+REGISTRY = EmbeddingFunctionRegistry()
+
+
+class EmbeddingFunctionModel(BaseModel, ABC):
+    """
+    A callable ABC for embedding functions
+    """
+
+    source_column: Optional[str]
+    vector_column: str
+
+    @abstractmethod
+    def __call__(self, *args, **kwargs):
+        pass
+
+
+@REGISTRY.register()
+class SentenceTransformerEmbeddingFunction(EmbeddingFunctionModel):
+    name: str = "all-MiniLM-L6-v2"
+    device: str = "cpu"
+    normalize: bool = False
+
+    @property
+    def embedding_model(self):
+        """
+        Get the sentence-transformers embedding model specified by the
+        name and device. This is cached so that the model is only loaded
+        once per process.
+        """
+        return self.__class__.get_embedding_model(self.name, self.device)
+
+    def __call__(self, texts: Union[str, List[str]]) -> List[np.array]:
+        """
+        Get the embeddings for the given texts
+
+        Parameters
+        ----------
+        texts: str or list[str]
+            The texts to embed
+        """
+        if isinstance(texts, str):
+            texts = [texts]
+        elif isinstance(texts, pa.Array):
+            texts = texts.to_pylist()
+        elif isinstance(texts, pa.ChunkedArray):
+            texts = texts.combine_chunks().to_pylist()
+        return self.embedding_model.encode(
+            list(texts),
+            convert_to_numpy=True,
+            normalize_embeddings=self.normalize,
+        ).tolist()
+
+    @classmethod
+    @cached(cache={})
```

Comment:
How big is a sentence transformer model?  Do we need to file a future issue to maybe control / monitor the size of this cache?

Reply:
the default is ~90mb in pickle format (https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/tree/main) and roughly 150mb when loaded into memory. I added a TODO already in the docstring "TODO: use lru_cache instead with a reasonable/configurable maxsize". I can file a Github issue as well if you like